### PR TITLE
Support for null strings

### PR DESCRIPTION
**Pull request by jitsni**
* it is encoded as "0xff" (-1 as the length). Hence, a String with length=255 is
not allowed

https://github.com/reaktivity/nukleus-maven-plugin/issues/34

### CODE REVIEW

**Comment by jitsni** [comment]
public StringFW build() {
      if (!valueSet) {
        set("", StandardCharsets.UTF_8);
      }
      return super.build();
    }

Should we call with set(null, UTF_8) in the above ?

**Comment by jfallows** [comment]
```
public StringFW build() {
  if (!valueSet) {
    set(null, StandardCharsets.UTF_8);
  }
  return super.build();
}
```

Having `null` as the default default-value seems better than empty-string so we can more easily distinguish the case where empty-string is explicit.

File: src/main/java/org/reaktivity/nukleus/maven/plugin/internal/generate/String16FlyweightGenerator.java [hunk 1]
```diff
@@ -235,12 +246,24 @@ private MethodSpec setStringMethod()
...
+                    .addStatement("throw new IllegalArgumentException("+
+                            "String.format(\"Value '%s' too big for String field (length < 65535)\", value))")
+                    .endControlFlow()
                     .addStatement("MutableDirectBuffer buffer = buffer()")
                     .addStatement("int offset = offset()")
                     .addStatement("int newLimit = offset + charBytes.length + FIELD_SIZE_LENGTH")
                     .addStatement("checkLimit(newLimit, maxLimit())")
-                    .addStatement("buffer.putShort(offset, (short) charBytes.length)")
+                    .addStatement("buffer.putShort(offset, (short) length)")
                     .addStatement("buffer.putBytes(offset + 2, charBytes)")
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/maven/plugin/internal/generate/String16FlyweightGenerator.java]: Move `if` condition to simplify code, and also eliminate unneeded garbage via `new byte[0]`.

```
return methodBuilder("set")
.addModifiers(PUBLIC)
.returns(stringType.nestedClass("Builder"))
.addParameter(String.class, "value")
.addParameter(Charset.class, "charset")
.beginControlFlow("if (value == null)")
.addStatement("int newLimit = offset + FIELD_SIZE_LENGTH")
.addStatement("checkLimit(newLimit, maxLimit())")
.addStatement("buffer.putShort(offset, (short) -1)")
.addStatement("limit(newLimit)")
.nextControlFlow("else")
.addStatement("byte[] charBytes = value.getBytes(charset)")
.addStatement("int newLimit = offset + FIELD_SIZE_LENGTH + charBytes.length")
.addStatement("checkLimit(newLimit, maxLimit())")
.addStatement("buffer.putShort(offset, (short) charBytes.length)")
.addStatement("buffer.putBytes(offset + 2, charBytes)")
.addStatement("limit(newLimit)")
.endControlFlow()
.addStatement("valueSet = true")
.addStatement("return this")
.build();
```

File: src/main/java/org/reaktivity/nukleus/maven/plugin/internal/generate/String16FlyweightGenerator.java [hunk 2]
```diff
@@ -235,14 +240,23 @@ private MethodSpec setStringMethod()
...
+                    .beginControlFlow("if (value == null)")
+                    .addStatement("int newLimit = offset() + FIELD_SIZE_LENGTH")
+                    .addStatement("checkLimit(newLimit, maxLimit())")
+                    .addStatement("buffer().putShort(offset(), (short) -1)")
+                    .addStatement("limit(newLimit)")
+                    .nextControlFlow("else")
+                    .beginControlFlow("if (value.length() >= 65535)")
+                    .addStatement("throw new IllegalArgumentException("+
+                            "String.format(\"Value '%s' too big for String field (length < 65535)\", value))")
+                    .endControlFlow()
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/maven/plugin/internal/generate/String16FlyweightGenerator.java]: Suggest generating `checkLength(int)` method with implementation:

```
private static void checkLength(
    int length)
{
    final int maxLength = 65534;
    if (length > maxLength)
    {
        final String msg = String.format("length=%d is beyond maximum length=%d", length, maxLength);
        throw new IndexOutOfBoundsException(msg);
    }
}
```

then call `checkLength(charBytes.length)` at line 254 below.

Note: there is a bug for multi-byte character UTF-8 strings with this current code, as `value.length()` is not the same as `charBytes.length` for multi-byte characters.

File: src/main/java/org/reaktivity/nukleus/maven/plugin/internal/generate/String16FlyweightGenerator.java [hunk 3]
```diff
@@ -262,6 +257,20 @@ private MethodSpec setStringMethod()
                     .build();
         }
 
+        private MethodSpec checkLengthMethod()
+        {
+            return methodBuilder("checkLength")
+                    .addModifiers(PRIVATE)
```
Comment (jfallows) [inline src/main/java/org/reaktivity/nukleus/maven/plugin/internal/generate/String16FlyweightGenerator.java]: `STATIC` needed

**Pull request merged**